File: coffeetime-parent/coffeetime-scala/src/main/scala/org/mrseasons/coffeetime/scala/_11_class/Man.scala
package org.mrseasons.coffeetime.scala._12_class

/**
 * Created by mrseasons on 2015/2/12.
 */
//  main constructor
class Man(val name: String, val age: Int, private var from: String = "USA", description: String = "none") {

  println("sentences in main constructor")

  //  slave constructor
  def this(name: String) {
    //  delegate to the main constructor
    this(name, 0)
  }

  def this(age: Int) {
    this("Default Name", age)
  }

  def this() {
    this("Default Name")
  }

}

class Woman private(val name: String, val age: Int) {
  def this(name: String) {
    this(name, 0)
  }
}

File: coffeetime-parent/coffeetime-kotlin/src/main/kotlin/org/mrseasons/coffeetime/kotlin/_11_class/Man.kt
package org.mrseasons.coffeetime.kotlin._11_class

/**
 * Created by mrseasons on 2015/06/05.
 */
//  primary constructor
//  var for mutable and val for read-only
class Man(val name: String, val age: Int, private var from: String = "USA", description: String = "none") {

    init {
        println("sentences in primary constructor")
    }

    //  slave constructor
    //  delegate to the primary constructor
    constructor(name: String) : this(name, 0) {
    }

    constructor(age: Int) : this("Default Name", age) {
    }

    constructor() : this("Default Name") {
    }
}

class Woman private constructor(val name: String, val age: Int) {
    constructor(name: String) : this(name, 0) {

    }
}